src/maxtree.jl

```diff
@@ -0,0 +1,881 @@
+
+# Gets the path to the root of the max-tree (defined by `ancestors`) starting from
+# `node`. The indices of the nodes along the path are stored into `path`.
+# Returns the index of the root
+@inline function rootpath!(path::Vector{Int},
+                           ancestors::AbstractArray{<:Integer},
+                           node::Integer)
+    empty!(path)
+    p = node
+    @label next
+    push!(path, p)
+    @inbounds q = ancestors[p]
+    if p != q
+        p = q
+        @goto next
+    end
+    return p
+end
+
+# Corrects `maxtree.parents`, so that the parent of every pixel
+# is a canonical node. I.e. for each connected component at a level l,
+# all pixels point to the same representative, which in turn points to the
+# representative pixel at the next level.
+#
+# A canonical node ``p`` is either a root of the max-tree, or
+# ``image(parent(p)) != image(p)``
+function canonize!(maxtree::MaxTree{N}, image::AbstractArray{<:Any, N}) where N
+    @assert size(maxtree) == size(image)
+    parents = maxtree.parents
+    @inbounds for p in maxtree.traverse
+        q = parents[p]
+        if image[q] == image[parents[q]]
+            parents[p] = parents[q]
+        end
+    end
+    return maxtree
+end
+
+# Checks whether the `pixel` + `offset` is still a valid pixel
+# contained inside the image area defined by `axes`.
+@generated isvalid_offset(offset::NTuple{N, Int}, pixel::CartesianIndex{N},
+                          axes::NTuple{N}) where N =
+    quote
+        Base.Cartesian.@nall $N i -> in(pixel[i] + offset[i], axes[i])
+    end
+
+"""
+    rebuild!(maxtree::MaxTree, image::AbstractArray,
+             neighbors::AbstractVector{NTuple}) -> maxtree
+
+Rebuilds the `maxtree` for the `image` using `neighbors` as the pixel
+connectivity specification.
+
+# Arguments
+- `neighbors`: the vector of the offsets (``o_i``) to the cartesian index of
+   the pixel (``p``), so that ``p + o_i`` is the cartesian index of the ``i``-th
+   neighbor of ``p``.
+
+# Details
+The pixels in the connected components generated by the method should be
+connected to each other by a path through neighboring pixels (as defined by
+`neighbors`).
+
+# See also
+[`MaxTree`](@ref)
+"""
+function rebuild!(maxtree::MaxTree{N},
+                  image::AbstractArray{<:Number, N},
+                  #=mask::AbstractArray{Bool, N},=#
+                  neighbors::AbstractVector{NTuple{N, Int}}) where N
+    # pixels need to be sorted according to their gray level.
+    sortperm!(maxtree.traverse, vec(image), rev=maxtree.rev)
+    # some sufficiently inner point
+    center = CartesianIndex{N}(ntuple(i -> 1 + size(image, i) ÷ 2, Val{N}()))
+    # offsets for linear indices (note that they are sorted in ascending order)
+    neighbor_offsets = getindex.(Ref(LinearIndices(image)),
+        [CartesianIndex{N}(ntuple(i -> center[i] + n[i], Val{N}())) for n in neighbors]) .-
+        getindex(LinearIndices(image), center)
```

Comment:
+1 for making the code more clear. Will do in the revision.
`Ref` is to avoid broadcasting `getindex()` over `LinearIndices(image)` (it should only be broadcasted over the neighbor cartesian indices).